Mathlib/FieldTheory/PrimitiveElement.lean

```diff
@@ -214,6 +251,120 @@ noncomputable def powerBasisOfFiniteOfSeparable : PowerBasis F E :=
+    show finrank K (⊤ : IntermediateField K E) = finrank K E from finrank_top K E,
+    ← natDegree_map_eq_of_injective (NoZeroSMulDivisors.algebraMap_injective K E) (minpoly K α),
+    show (minpoly K α).map (algebraMap K E) = g from rfl,
+    ← hp,
+    natDegree_map_eq_of_injective (NoZeroSMulDivisors.algebraMap_injective K'' E),
+    hqdvdp,
+    natDegree_mul (left_ne_zero_of_mul hpne0) (right_ne_zero_of_mul hpne0),
+    ← IntermediateField.adjoin.finrank (K := K'') (x := α) (hx := isIntegral_of_finite K'' α),
+    finite_intermediateField_of_exists_primitive_element.aux_1 F E _ hprim K'',
+    show finrank K'' (⊤ : IntermediateField K'' E) = finrank K'' E from finrank_top K'' E] at hpdeg
+  replace hpdeg : finrank K E ≥ finrank K'' E := by linarith only [hpdeg, Nat.le_add_right]
+  exact (eq_of_le_of_finrank_le' inf_le_left hpdeg).symm
+
+-- A finite simple extension has only finitely many intermediate fields
+theorem finite_intermediateField_of_exists_primitive_element
+    (h : ∃ α : E, F⟮α⟯ = ⊤) : Finite (IntermediateField F E) := by
+  obtain ⟨α, hprim⟩ := h
+  -- Let `f` be the minimal polynomial of `α ∈ E` over `F`
+  let f : Polynomial F := minpoly F α
+  let G := { g : Polynomial E // g.Monic ∧ g ∣ f.map (algebraMap F E) }
+  -- Then `f` has only finitely many monic factors
+  have hfin : Finite G := @Finite.of_fintype _ $ fintypeSubtypeMonicDvd
+    (f.map (algebraMap F E)) $ by
+      rw [Polynomial.map_ne_zero_iff $ NoZeroSMulDivisors.algebraMap_injective F E]
+      exact minpoly.ne_zero_of_finite F α
+  -- If `K` is an intermediate field of `E/F`, let `g` be the minimal polynomial of `α` over `K`
+  -- which is a monic factor of `f`
+  let g : IntermediateField F E → G := fun K =>
+    ⟨(minpoly K α).map (algebraMap K E), ⟨Monic.map (algebraMap K E) $
+      minpoly.monic $ isAlgebraic_iff_isIntegral.1 $ isAlgebraic_of_finite K α,
+    by
+      obtain ⟨fac, hfac⟩ := minpoly.dvd_map_of_isScalarTower F K α
+      use fac.map (algebraMap K E)
+      apply_fun Polynomial.map (algebraMap K E) at hfac
+      rw [← Polynomial.map_mul, ← hfac, Polynomial.map_map]
+      congr 1⟩⟩
+  -- The map `K ↦ g` is injective
+  have hinj : Function.Injective g := fun K K' heq => by
+    rw [Subtype.mk.injEq] at heq
+    have h := Field.finite_intermediateField_of_exists_primitive_element.aux_2
+      F E α hprim K K' heq
+    rwa [inf_comm, ← Field.finite_intermediateField_of_exists_primitive_element.aux_2
+      F E α hprim K' K heq.symm] at h
+  -- Therefore there are only finitely many intermediate fields
+  exact Finite.of_injective g hinj
+
+/-- **Steinitz theorem**: a finite extension `E` of `F` has a
+  primitive element (i.e. there is an `α ∈ E` such that `F⟮α⟯ = (⊤ : Subalgebra F E)`)
+  if and only if there exist only finitely many intermediate fields between `E` and `F`. -/
+theorem exists_primitive_element_iff_finite_intermediateField :
+    (∃ α : E, F⟮α⟯ = ⊤) ↔ Finite (IntermediateField F E) :=
```

Comment:
The forward direction also holds if `E/F` is only algebraic (with finite dimensionality implied by the left condition), and the backward direction should hold for any extension `E/F`, with algebraicity and finite dimensionality implied by the right condition.

Reply:
It complains about `[Algebra.IsAlgebraic F E]` because `Algebra.IsAlgebraic` is not a class. So I have to write `(halg : Algebra.IsAlgebraic F E)`.

Reply:
Yeah, indeed you have to do that when it's not a class. Having a FiniteDimensional version might be more convenient, but I doubt this theorem will be used very often ...